base/reinterpretarray.jl

```diff
@@ -0,0 +1,67 @@
+"""
+Gives a reinterpreted view (of element type T) of the underlying array (of element type S).
+If the size of `T` differs from the size of `S`, the array will be compressed/expanded in
+the first dimension.
+"""
+struct ReinterpretArray{T,S,N,A<:AbstractArray{S, N}} <: AbstractArray{T, N}
+    parent::A
+    Base.reinterpret(::Type{T}, a::A) where {T,S,N,A<:AbstractArray{S, N}} = new{T, S, N, A}(a)
+end
+
+Base.eltype(a::ReinterpretArray{T}) where {T} = T
+function Base.size(a::ReinterpretArray{T,S}) where {T,S}
+    psize = size(a.parent)
+    if sizeof(T) > sizeof(S)
+        size1 = div(psize[1], div(sizeof(T), sizeof(S)))
+    else
+        size1 = psize[1] * div(sizeof(S), sizeof(T))
+    end
+    tuple(size1, Base.tail(psize)...)
+end
+
+@inline @propagate_inbounds function Base.getindex(a::ReinterpretArray{T,S,N}, inds::Vararg{Int, N}) where {T,S,N}
+    if sizeof(T) == sizeof(S)
+        return reinterpret(T, a[inds...])
+    elseif sizeof(T) > sizeof(S)
+        nels = div(sizeof(T), sizeof(S))
+        ind_off = (inds[1]-1) * nels
```

Comment:
not inbounds if called as `ReinterpretArray(a)[]`

Reply:
maybe define that method specifically? `Base.getindex(a::ReinterpretArray) = a[1]`, if that's right?